launch/launch/actions/execute_process.py

```diff
@@ -0,0 +1,473 @@
+# Copyright 2018 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module for the ExecuteProcess action."""
+
+import asyncio
+import logging
+import os
+import shlex
+import signal
+import threading
+import traceback
+from typing import Any
+from typing import cast
+from typing import Dict
+from typing import Iterable
+from typing import List
+from typing import Optional
+from typing import Text
+
+from osrf_pycommon.process_utils import async_execute_process
+from osrf_pycommon.process_utils import AsyncSubprocessProtocol
+
+from .emit_event import EmitEvent
+from .opaque_function import OpaqueFunction
+from .timer_action import TimerAction
+from ..action import Action
+from ..event import Event
+from ..event_handler import EventHandler
+from ..event_handlers import OnShutdown
+from ..events import Shutdown
+from ..events.process import matches_action
+from ..events.process import ProcessExited
+from ..events.process import ProcessStarted
+from ..events.process import ProcessStderr
+from ..events.process import ProcessStdin
+from ..events.process import ProcessStdout
+from ..events.process import ShutdownProcess
+from ..events.process import SignalProcess
+from ..launch_context import LaunchContext
+from ..launch_description import LaunchDescription
+from ..some_actions_type import SomeActionsType
+from ..some_substitutions_type import SomeSubstitutionsType
+from ..substitution import Substitution
+from ..substitutions import LaunchConfiguration
+from ..substitutions import PythonExpression
+from ..utilities import create_future
+from ..utilities import is_a_subclass
+from ..utilities import normalize_to_list_of_substitutions
+from ..utilities import perform_substitutions
+
+_logger = logging.getLogger(name='launch')
+
+_global_process_counter_lock = threading.Lock()
+_global_process_counter = 0  # in Python3, this number is unbounded (no rollover)
+
+
+class ExecuteProcess(Action):
+    """Action that begins executing a process and sets up event handlers for the process."""
+
+    def __init__(
+        self, *,
+        cmd: Iterable[SomeSubstitutionsType],
+        cwd: Optional[SomeSubstitutionsType] = None,
+        env: Optional[Dict[SomeSubstitutionsType, SomeSubstitutionsType]] = None,
+        shell: bool = False,
+        sigterm_timeout: SomeSubstitutionsType = LaunchConfiguration('sigterm_timeout', default=5),
```

Comment:
I imagine you are setting the default via a LaunchConfiguration in order to retrieve its value when it's actually used (as opposed to just passing in a float). But is it actually possible that someone could modify it before then, externally? FWIU [it gets evaluated here](https://github.com/ros2/launch/pull/74/files#diff-61622adb9c2241797a64d96e6065c414R273) and we don't modify it internally, so maybe we can just pass it as a float?

Reply:
answering myself: it is modifiable externally in the same way [the `launch-prefix` configuration is](https://github.com/ros2/launch/pull/74/files#diff-61622adb9c2241797a64d96e6065c414R163).

Reply:
> I imagine you are setting the default via a LaunchConfiguration in order to retrieve its value when it's actually used (as opposed to just passing in a float).

Yes.

> But is it actually possible that someone could modify it before then, externally?

Yes, consider:

```python
ld = LaunchDescription([
  SetLaunchConfiguration('sigterm_timeout', 42.0),
  ExecuteProcess(cmd=['ls']),
  ExecuteProcess(
    cmd=['ls'],
    sigterm_timeout=EnvironmentVariable('CUSTOM_SIGTERM_TIMEOUT')),
  ExecuteProcess(cmd=['ls', '-la']),
  # ExecuteProcess(cmd=['ls', '-la'], sigterm_timeout=1.0),
])
```

The first and last execute processes would use the timeout set in the launch configuration, but the second would not because it was explicitly passed and the fourth (commented out) would be a nice to have interface where `sigterm_timeout` could also take a non-text/non-substitution numeric, but it's not necessary.

> FWIU it gets evaluated here and we don't modify it internally, so maybe we can just pass it as a float?

Not exactly, the substitutions that make up `sigterm_timeout` are passed to the `TimerAction` there, but they are not substituted until the `TimeAction` is visited, and the substitutions should not be evaluated until then I think. We couldn't expand the substitution there even if we wanted to, because when called via describe, it would not have a `LaunchContext` which is required to perform substitutions.

Reply:
yep, it doesn't get evaluated until the TimerAction is visited/executed: https://github.com/ros2/launch/pull/74/files#diff-a40b801740c81b57ba0a59a56cc0d36fR79.

I documented this behaviour in f3417d4